lib/python/treadmill_aws/garbage_collector.py

```diff
@@ -0,0 +1,62 @@
+"""AWS specific treadmill garbage collector
+"""
+
+import logging
+
+import jmespath
+
+from treadmill import admin
+from treadmill import context
+from treadmill_aws.awscontext import GLOBAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _ec2_instances():
+    instance_name = jmespath.compile(
+        "Reservations[].Instances[].[Tags[?Key=='Name'].Value][][]")
+
+    _LOGGER.info('Fetching valid instances from AWS')
+    instances = instance_name.search(GLOBAL.ec2.describe_instances())
+    _LOGGER.debug("%d valid instances found", len(instances))
+
+    return set(instances)
+
+
+class LDAP:
+    """LDAP garbage collection"""
+    client = admin.Server(context.GLOBAL.ldap.conn)
+
+    @staticmethod
+    def list():
+        """List LDAP server records that are not valid ec2 instances"""
+        ldap_servers = {host.get("_id") for host in LDAP.client.list({})}
+        return ldap_servers - _ec2_instances()
+
+    @staticmethod
+    def delete(hostname):
+        """Delete a LDAP server record"""
+        _LOGGER.info('removing %s from LDAP', hostname)
+        LDAP.client.delete(hostname)
+
+
+class IPA:
+    """IPA garbage collection"""
+    client = GLOBAL.ipaclient
+
+    @staticmethod
+    def list():
+        """List IPA server records that are not valid ec2 instances"""
+        return set(LDAP.client.get_hosts()) - _ec2_instances()
```

Comment:
idea is that list returns all servers known to the plugin, not the diff.

it is actuall wrong to create it as class variable, as connection is lazy.

return admin.Server(context.GLOBAL.ldap.conn).get_hosts()

(i did not realized there is get_hosts() method... )